### PR TITLE
Optimize policy compilation for N-of-N tresholds

### PR DESCRIPTION
**Pull request by darosior**
This adds a special case to the compilation of the `thresh(k, subs)` policy to
miniscript: when all sub-policies are keys and k == subs.len() this
compiles to a combination of and()s instead of a thresh().

The first commit adds some tests for the policy->miniscript compilation of different combinations of `k` and pubkeys under the `thresh()` policy.

The second commit special cases `k == pubkeys.len()` (when `> 20`) to eventually produce slightly more efficient Bitcoin Script:
```
<pubkey1> CHECKSIGVERIFY <pubkey2> CHECKSIGVERIFY  <pubkey3> CHECKSIGVERIFY ... <pubkeyN> CHECKSIG
```
Instead of
```
<pubkey1> CHECKSIG SWAP <pubkey2> CHECKSIG ADD SWAP  <pubkey3> CHECKSIG ADD SWAP ... <pubkeyN> CHECKSIG ADD N EQUAL
```

As mentioned by sipa on IRC we can also special-case `k==1` to a combination of `or()`s, but I left a `FIXME` as I intend to do this in a follow-up (can also push a commit here, as you prefer).

### CODE REVIEW

File: src/policy/compiler.rs
Comment (darosior) [inline src/policy/compiler.rs:1425]: This test is a bit hacky but I had no idea how to test it an other way as the compilation is not deterministic. Open to suggestion

File: src/policy/compiler.rs
Comment (sanket1729) [inline src/policy/compiler.rs:1001]: I think this is not special to keys, The else if code can be moved to match thresh rule. Transforming the 

Concrete::thresh(k, ref subs) into chain of `Ands` if `k == n`. That way this will also work for things that are not keys and for n < 20 too.

File: src/policy/compiler.rs
Comment (sanket1729) [reply src/policy/compiler.rs:1425]: We don't try to match the exact Miniscripts generated by the compilations. We only check whether the expected satisfaction costs are the same.

**Comment by sanket1729** [comment]
The complexity of the compiler is proportional to the number of nested or's. I am a bit concerned in transforming `Thresh(1, subs)` into a chain of `or`s for higher value of `k`. cc @apoelstra

File: src/policy/compiler.rs
Comment (darosior) [reply src/policy/compiler.rs:1001]: Hmm, I'm not sure it's efficient for `n < 20` ?

If using `CHECKMULTISIG` you have `4` bytes taken by OP codes (`n` pushed 2 times, `CHECKMULTISIG`, and the empty byte array), while you have `n` bytes taken by OP codes if using a set of `CHECKSIGVERIFY`s.

So, yeah, it's worth it for `n < 4 || n > 20`, will change that.

Comment (sanket1729) [reply src/policy/compiler.rs:1001]: I mean the there can be other `subs` apart from keys that can benefit from chaining `And` compilation. It is most likely the case that `thresh(1,sha256(H),pk(C),pk(C),pk(C))` should be treated as `and(sha256(H),and(pk(C),and(pk(C),pk(C))` would produce a better compilation

File: src/policy/compiler.rs
Comment (darosior) [reply src/policy/compiler.rs]: But that's what other tests do ? (I mostly used `compile_misc()` as an example)

Ok so I kept the miniscript assertion for those that compiled straight (thresh => multi and thresh => thresh), and added script_size assertion on all.

**Comment by darosior** [comment]
Amended to not use inclusive range syntax..

File: src/policy/compiler.rs
Comment (darosior) [reply src/policy/compiler.rs]: Ok, so we want this for policies where `k == subs.len()` *but not* when all `subs` are keys and `k in 4..=20` (in which case we want `multi`) ?

Comment (darosior) [reply src/policy/compiler.rs]: Finally went for:
- All keys and less than 20 keys ?
  - Yes => `multi`
  - No
    - => `k == subs.len()` ?
      - Yes => Nested `and()`s
      - No => `thresh()`

File: src/policy/compiler.rs
Comment (apoelstra) [inline src/policy/compiler.rs:1415]: What does "doesn't work" mean?

File: src/policy/compiler.rs
Comment (apoelstra) [inline src/policy/compiler.rs:1006]: It is probably better to compare the `best_compilations` returned here to the ones we computed using the threshold logic, and actually make sure that we're taking the best ones. It seems intuitively true that nested `and`s will beat `thresh` every time but I'm not sure.

Comment (sanket1729) [reply src/policy/compiler.rs:1006]: Agreed that `best_compilations` makes more sense here. But the current code always produces the most optimal compilation. Consider a particular compilation with all AND's replaced as BOOLAND. 
We would have 

`[E] [W] BOOLAND [W] BOOLAND .....`

The corresponding thresh would be:
`[E] [W] ADD [W] ADD [W] ADD.... k EQUAL`. 

Therefore, there exists at least one compilation that is better than the `thresh` one. In any case, this deserves a comment explaining this scenario.

File: src/policy/compiler.rs
Comment (sanket1729) [reply src/policy/compiler.rs:1415]: I know you saw this issue. But for completeness of future viewers( refer #114 )

Comment (darosior) [reply src/policy/compiler.rs:1415]: Will link the issue instead, and maybe remove the "FIXME" as it seems we're not going to fix it ?

Comment (sanket1729) [reply src/policy/compiler.rs:1415]: Yes, that would be great.

File: src/policy/compiler.rs
Comment (darosior) [reply src/policy/compiler.rs:1006]: I've expanded a bit more on the comment just above (was not sure how to phrase it though):
> // Not a threshold, it's always more optimal to translate it to and()s as we save the
> // resulting threshold check (N EQUAL) in any case.

**Comment by sanket1729** [comment]
tAck 688b0a84673f52e463f77dc1338e786da2fb1742 . Will merge once #118 is merged so that the current invariant fuzz tests
`policy -> compile -> lift == policy -> lift` is preserved.

File: src/policy/compiler.rs
Comment (apoelstra) [reply src/policy/compiler.rs:1007]: I think this phrasing is fine

**Review by apoelstra — APPROVED** [review]
Neat, lgtm. I'm not sure we want a "FIXME" comment for a future optimization we're unsure how to do, but we can address that later

**Comment by darosior** [comment]
Will open an issue regarding the uncertainty of fixing the FIXME, this'll end up with either me implementing it or removing the FIXME.

**Pull request merged**
